fix(avm)!: Append-only tree reads

MirandaWood opened the pull request:
### AVM append-only tree reads
Closes #17330

This PR unifies the behaviour between AVM simulation and circuits for reading leaves in append-only trees (note hash and l1 to l2 message trees). There are two key changes:

- Reading beyond the maximum index a tree can hold now 'throws' in both cases and is handled by setting a `false` response to the opcodes `NOTEHASHEXISTS` & `L1TOL2MSGEXISTS`
  - We now just skip the world state call by throwing in ts, mirroring circuit behaviour where we skip the tree lookup if `note_hash_leaf_in_range == 0`
  - I additionally added changes in the cpp append-only tree to fix an issue where `get_leaf` would return the leaf at index `i` when calling `get_leaf(MAX_INDEX + i)` (see #17755) and instead return null
 - Reading an unwritten leaf (i.e. at some valid index `i < MAX_INDEX`) now returns an empty leaf instead of null
   - This required changes in cpp `world_state` to ensure we differentiate between responses where a leaf is not yet written to vs a leaf which cannot exist at an invalid index

### TODOs/Notes

 - The handling of invalid indices in the cpp append-only tree is a bit of a quick fix, and I'm not sure the best way to handle, see #17755
 - Handle `leaf_index > blockData.size` - ~I'm not sure whether the intended behaviour here is to 'fail' (current impl - returns with `response.success = false`) or mirror the behaviour of `get_leaf` without specifying a block, which is to return an empty leaf~ Keeping with `response.success = false` since note and msg trees don't seem to ever reach this path (and changing the behaviour may cause far reaching problems with other trees specifying block number in the revision e.g. archive tree).

MirandaWood commented on barretenberg/cpp/src/barretenberg/crypto/merkle_tree/append_only_tree/content_addressed_append_only_tree.hpp:
```diff
@@ -714,12 +730,24 @@ void ContentAddressedAppendOnlyTree<Store, HashingPolicy>::get_leaf(const index_
                                               " for block ",
                                               blockNumber,
-                                              ", leaf index out of range.");
+                                              ", leaf index out of block range.");
```
See description - I'm not sure what `blockData.size` really means in this context and whether we should return a 0 leaf?
This path isn't reached by the AVM simulation as far as I can see.

dbanks12 replied:
@alexghr could you advise here?

MirandaWood replied:
Update: investigating if/how this path is ever reached by AVM simulation (I initially could not find any examples, but it may be possible), in which case this would need to return a 0 leaf as well to match circuit behaviour

Update update: seems like note and msg trees will not reach this path, so don't require a 0 leaf rather than a failure. Keeping it as a failure to prevent problems with other trees (e.g. archive tree does reach this path)

alexghr commented on yarn-project/simulator/src/public/state_manager/state_manager.ts:
```diff
@@ -185,12 +185,25 @@ export class PublicPersistableStateManager {
+    } catch (error) {
+      // If the index is out of range, note_hash_leaf_in_range = 0 and the circuit returns false:
+      if (error instanceof Error && error.message.includes('out of range')) {
+        return Promise.resolve(false);
```
this feels a bit brittle. Could we use a custom Error subclass?

MirandaWood replied:
good point, done!

dbanks12 commented on yarn-project/simulator/src/public/state_manager/state_manager.ts:
```diff
@@ -185,12 +190,25 @@ export class PublicPersistableStateManager {
+      if (gotLeafValue == undefined) {
+        // TODO(MW): Move to public_db_sources? We should no longer get unhandled null responses from WS
+        throw new Error('Unexpected null response from getNoteHash');
+      }
```
Yeah, maybe just turn this into an `assert` in public_db_sources?

dbanks12 commented on yarn-project/simulator/src/public/state_manager/state_manager.ts:
```diff
@@ -290,12 +308,25 @@ export class PublicPersistableStateManager {
+        // TODO(MW): Move to public_db_sources? We should no longer get unhandled null responses from WS
+        throw new Error('Unexpected null response from getL1ToL2LeafValue');
+      }
+      const exists = valueAtIndex.equals(msgHash);
```
Same as above. Assert in public db sources?

dbanks12 commented on yarn-project/simulator/src/public/side_effect_errors.ts:
```diff
@@ -59,3 +61,36 @@ export class NullifierCollisionError extends SideEffectError {
     this.name = 'NullifierCollisionError';
   }
 }
+
+/**
+ * Any error that can be thrown during side effect reads in public.
+ * Note: NOT considered CheckedPublicExecutionErrors because they do not result in reverts
+ * in the circuit. Currently only includes append-only tree reads.
+ */
+export abstract class SideEffectReadError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'SideEffectReadError';
+  }
+}
+
+export class IndexOutOfRangeError extends SideEffectReadError {
+  constructor(tree: string, index: number, limit: number) {
+    super(`Attempting to read index ${index} of ${tree} tree with maximum ${limit} leaves`);
+    this.name = 'IndexOutOfRangeError';
+  }
+}
+
+export class NoteHashIndexOutOfRangeError extends IndexOutOfRangeError {
+  constructor(index: number) {
+    super('note hash', index, NOTE_HASH_TREE_LEAF_COUNT);
+    this.name = 'NoteHashIndexOutOfRangeError';
+  }
+}
+
+export class L1ToL2MessageIndexOutOfRangeError extends IndexOutOfRangeError {
+  constructor(index: number) {
+    super('l1 to l2 message', index, L1_TO_L2_MSG_TREE_LEAF_COUNT);
+    this.name = 'L1ToL2MessageIndexOutOfRangeError';
+  }
+}
```
Cool! For the comment, I wouldn't say "not result in reverts in the circuit". The real reason they are not "checked" errors is because they are really internal errors that should be caught by the state manager and should never be known/checked by the AvmSimulator or PublicTxSimulator.

MirandaWood replied:
Makes sense, will update! My wording here was more trying to point out that other errors in this file result in the column `reverted = 1`, but these new errors result in something like `note_hash_leaf_in_range = 0` which just skips the lookups and sets a (valid) result of 0. But I think I failed in that regard 😅

dbanks12 reviewed (approved):
LGTM, but just a few nits. And let's get confirmation about one of your TODOs (blockData.size)

MirandaWood committed:
fix(avm)!: Append-only tree reads

### AVM append-only tree reads
Closes #17330

This PR unifies the behaviour between AVM simulation and circuits for reading leaves in append-only trees (note hash and l1 to l2 message trees). There are two key changes:

- Reading beyond the maximum index a tree can hold now 'throws' in both cases and is handled by setting a `false` response to the opcodes `NOTEHASHEXISTS` & `L1TOL2MSGEXISTS`
  - We now just skip the world state call by throwing in ts, mirroring circuit behaviour where we skip the tree lookup if `note_hash_leaf_in_range == 0`
  - I additionally added changes in the cpp append-only tree to fix an issue where `get_leaf` would return the leaf at index `i` when calling `get_leaf(MAX_INDEX + i)` (see #17755) and instead return null
 - Reading an unwritten leaf (i.e. at some valid index `i < MAX_INDEX`) now returns an empty leaf instead of null
   - This required changes in cpp `world_state` to ensure we differentiate between responses where a leaf is not yet written to vs a leaf which cannot exist at an invalid index

### TODOs/Notes

 - The handling of invalid indices in the cpp append-only tree is a bit of a quick fix, and I'm not sure the best way to handle, see #17755
 - Handle `leaf_index > blockData.size` - ~I'm not sure whether the intended behaviour here is to 'fail' (current impl - returns with `response.success = false`) or mirror the behaviour of `get_leaf` without specifying a block, which is to return an empty leaf~ Keeping with `response.success = false` since note and msg trees don't seem to ever reach this path (and changing the behaviour may cause far reaching problems with other trees specifying block number in the revision e.g. archive tree).
